### PR TITLE
Make it a bit easier to debug TyKind::Error errors

### PR DESCRIPTION
**Pull request by mark-i-m**
r? @eddyb 

Not sure how good of an idea this is... but it occurred to me that we can do more to enforce the invariant that `TyKind::Error` should only happen if errors are going to be emitted. We could also use the function to due a delay bug for good measure.

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -841,6 +841,12 @@ impl<'tcx> CommonTypes<'tcx> {
             trait_object_dummy_self: mk(Infer(ty::FreshTy(0))),
         }
     }
+
+    #[inline]
+    pub fn err(&self) -> Ty<'tcx> {
```
Comment (eddyb) [inline src/librustc/ty/context.rs]: If you want to use `delay_span_bug`, here's a cool trick: add `#[track_caller]` to this function, then use `std::panic::Location::caller()` to get a location to print for the `delay_span_bug`.

cc @anp

**Comment by eddyb** [comment]
cc @rust-lang/compiler Let's get some visibility on this!

The other idea in this area is that we can make `ErrorReported` a token impossible to obtain without reporting an error (or using `delay_span_bug` I suppose), and then make `ty::Error` hold one, so that it acts as sort of a "proof".

**Comment by varkor** [comment]
> The other idea in this area is that we can make ErrorReported a token impossible to obtain without reporting an error

This is a really nice idea :+1:

**Comment by mark-i-m** [comment]
@eddyb In order for that to be meaningful, we would need to not intern `TyKind::Error` in `CommonTypes` as we do currently. Do you know if there would be any performance effects of this?

**Comment by mark-i-m** [comment]
I've pushed a rough prototype (not compiling yet, but I need to go to bed) with both of @eddyb's suggestions above. It's not clear that we need both, and between the two, I think prevention is better than detection, but I guess both can't hurt...

**Comment by eddyb** [comment]
I would call the variable `proof`, and have methods return it automatically without needing separate forms. Also, can we change `ErrorReported` in a separate PR, before using it in `Ty::Error`?

Btw there an issue for the `ErrorReported` thing: #69426.

**Comment by mark-i-m** [comment]
@eddyb 
> I would call the variable proof,

Which variable? I'm already using the name proof.

> and have methods return it automatically without needing separate forms

Do you mean the methods of `DiagnosticBuilder`? The challenge is that we should only return `ErrorProof` if an _error_ was emited, not just a warning or help.

> Also, can we change ErrorReported in a separate PR, before using it in Ty::Error?

I had created a new `ErrorProof` type for this purpose. The problem is that a lot of other code is using `ErrorReported` already to just indicate an error condition in the code somewhere, not necessarily as we want to use it now.

**Comment by eddyb** [comment]
> Which variable? I'm already using the name proof.

Ugh, I accidentally a word. I meant to say I would call variables `reported` or `error` instead of `proof`. (edited the original comment as well)

> The challenge is that we should only return ErrorProof if an *error* was emited, not just a warning or help.

Oh, is this about the "level" thing? Can we make that a const generic parameter of the `DiagnosticBuilder` or is that too extreme?

Ideally calling `.emit()` on a builder created by `struct_span_err!` should give `ErrorReported`.

> I had created a new `ErrorProof` type for this purpose. The problem is that a lot of other code is using `ErrorReported` already to just indicate an error condition in the code somewhere, not necessarily as we want to use it now.

Ah, whereas I would start by changing `ErrorReported`, making all of its current usecases guaranteed to have errors reported, before I ever touch `ty::Error`.
This is not a priority and I'd rather us get it right from the start.

**Comment by mark-i-m** [comment]
> Oh, is this about the "level" thing? Can we make that a const generic parameter of the DiagnosticBuilder or is that too extreme?
>
> Ideally calling .emit() on a builder created by struct_span_err! should give ErrorReported.

Hmm, I think I would rather not add a generic argument to all uses of `DiagnosticBuilder`. It seem like it would be infectious, but I might be wrong.

Another idea is to have something like:

```rust
enum ErrorReported {
  /// Emitted something with level < error
  Emitted,
  /// Emitted something with that will fail compilation
  EmittedWithProof(ErrorProof),
}

impl DiagnosticBuilder {
  pub fn emit(...) -> ErrorReported { ... }
}
```

I think this would accomodate both the current use cases and the ones where we want a proof.

**Comment by eddyb** [comment]
`ErrorReported` is *already* used like a proof token, it's not just *actually* one.
AFAIK it's never used anywhere compilation could still succeed, that's sort of the point.

**Comment by eddyb** [comment]
> Hmm, I think I would rather not add a generic argument to all uses of `DiagnosticBuilder`.

Hmm, actually, a type parameter would make sense, whereas const generics would be hard to use. Only lints need variable level.

And most error reporting code doesn't name `DiagnosticBuilder` by name so this seems fine.

**Comment by mark-i-m** [comment]
@eddyb so let me see if I understand correctly. Are you thinking of something like this:

```rust
// `true` if fails compilation
struct DiagnosticBuilder<const IS_ERROR: bool> {...}

impl DiagnosticBuilder<true> {
  pub fn emit() {
    // like today
  }
}

impl DiagnosticBuilder<false> {
  pub fn emit() -> Proof { ... } // returns proof
}

impl DiagnosticBuild<IS_ERROR> {
  // all the others remain generic
}
```

**Comment by mark-i-m** [comment]
> ErrorReported is already used like a proof token, it's not just actually one.
AFAIK it's never used anywhere compilation could still succeed, that's sort of the point.

I played around with this a bit yesterday, and I feel like I ran into cases where `ErrorReported` values were created in case they would be needed later. I can try to find them again if you want.

**Comment by eddyb** [comment]
Forget I ever mentioned `const` generics (which I was imagining `const LEVEL: Level`).

What I have in mind most recently is this:
```rust
// Could even make this not constructible without calling
// a method that registers something similar to delay_bug.
// But that's overkill for now.
pub struct AlwaysError;

mod sealed {
	use super::{AlwaysError, ErrorReported, Level};

    pub trait LevelTrait {
        type Emitted;
        fn emit(self) -> (Level, Self::Emitted);
    }

    impl LevelTrait for Level {
        type Emitted = ();
        fn emit(self) -> (Level, Self::Emitted) {
            (self, ())
        }
    }

    impl LevelTrait for AlwaysError {
        type Emitted = ErrorReported;
        fn emit(self) -> (Level, Self::Emitted) {
            (Level::Error, ErrorReported { private: () })
        }
    }
}

struct DiagnosticBuilder<L> {
    level: L,
    ...
}

impl<L: sealed::LevelTrait> DiagnosticBuilder<L> {
    pub fn emit(self) -> L::Emitted {
        ...
        let (level, emitted) = self.level.emit();
        ...
        emitted
    }
}
```

**Comment by mark-i-m** [comment]
Hmm... I could probably do something like that...

**Comment by mark-i-m** [comment]
How would one deal with cancellation of errors? Ideally one would need to return the proof to cancel an error.

**Comment by mark-i-m** [comment]
Hmm... ok this is getting very messy. There are too many weird edge cases. I still prefer the enum approach of https://github.com/rust-lang/rust/pull/70245#issuecomment-602284802 .

(Also, I think we can just forget about cancellation for now. Anyway, we can do something like track_callers above to help debug weird cases where an error is cancelled.)

**Comment by mark-i-m** [comment]
@eddyb after trying both, the enum approach is way easier to implement, largely because it doesn't clash with cancellation or lint levels; it just seems to be easier to do at runtime than compile time.

Also, I'm not sure what to do about a bunch of uses of `types.err` (please see the next rash of CI failures). Specifically, during type folding in a few places, we return `types.err` before the actual error is reported AFAICT, presumably to construct the error itself (e.g. src/librustc_infer/infer/resolve.rs:193)

**Comment by eddyb** [comment]
I still think it's a waste of time to try to change `ty::Error`, it's far more disruptive than making `ErrorReported` correct-by-default separately.
If you want I can attempt what I'm suggesting once I'm done with #70164.

As for error cancellation, we should try to avoid that as much as possible.
Presumably a cancellable error wouldn't guarantee `ErrorReported`.
Actually, does cancelling happen before `.emit()`? Because that would be fine.

> Specifically, during type folding in a few places, we return types.err before the actual error is reported AFAICT, presumably to construct the error itself (e.g. src/librustc_infer/infer/resolve.rs:193)

Ah, wait, inference contexts have snapshots, so you can try something and not actually report any errors. This might mean `ty::Error` can't hold any proof, or we'd need a different design.

**Comment by mark-i-m** [comment]
> As for error cancellation, we should try to avoid that as much as possible.
Presumably a cancellable error wouldn't guarantee ErrorReported.
Actually, does cancelling happen before .emit()? Because that would be fine.

AFAICT, a `Diagnostic` can be canceled at any time, even after emission. One idea is to make `ErrorReported` contain the `Diagnostic` itself. That way, if the `Diagnostic` is canceled, the `ErrorReported` must be revoked too.

----

Ok, for now, I'm going to pursue the `track_caller` idea you mentioned above, apart from the `ErrorReported` or `ty::Error`.

**Pull request closed without merging**